### PR TITLE
Int6 GPTQ-lite + LeakyReLU(0.5)^2 + EMA + 11L MLP3x

### PR DESCRIPTION
**Pull request by zeytx**
## Summary

- 11 transformer layers (vs 9 baseline) with 3x MLP expansion
- LeakyReLU(0.5)^2 activation (~0.003 BPB improvement over ReLU^2)
- Int6 per-row GPTQ-lite quantization with clip search + zstd-22 compression
- Late QAT via STE (activates when LR scale < 0.15)
- EMA weight averaging (decay 0.997)
- GQA: 8 query heads, 4 KV heads

## Results

- **val_bpb: 1.1807** (pre-quantization, 8xH100 SXM, 9881 steps in 10 min)
- Compressed model size: ~3.9 MB (well under 16MB limit)
- 26.5M parameters

## Run command

```bash
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

Requires `zstandard` package (`pip install zstandard`).

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Int6 GPTQ-lite + LeakyReLU(0.5)^2 + EMA + 11L MLP3x

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## PR #805 — Int6 + LeakyReLU(0.5)^2 + EMA + 11L MLP3x **Head SHA:** 6c2645449e52bf12365743b67c268edbeb3f9af8 **Files changed:** - `records/track_10min_16mb/2026-03-25_Int6_LeakyReLU2_EMA_11L_MLP3x/README.md` - `records/track_10min_16mb/2026-03-25_Int6_LeakyReLU2_EMA_11L_MLP3x/submission.json` - `records/track_10min_16mb/2026-03-25_Int6_LeakyReLU2_EMA_11L_MLP3x/train_gpt.py` --- ### Check 1: ILLEGAL n-gram family bug (target XOR'd into hash key / BigramHash) **CLEAR.** No bigram, n-gram, BigramHash, or hash-key construction anywhere in the file. No token XOR operations. Tokenizer LUTs (`base_bytes_lut`, `has_leading_space_lut`, `is_boundary_token_lut`) are used only for BPB byte-counting (lines 301-303, 327-328), not to influence logits or loss targets. ### Check 2: ILLEGAL Pre-Quant TTT (multi-epoch on val_tokens without score-first) **CLEAR.** `val_tokens` is loaded at line 905 and passed exclusively to `eval_val()` (lines 1087 and 1238). `eval_val()` runs entirely under `torch.inference_mode()` (lines 286, 311) — no gradients, no optimizer steps, no `.backward()` calls touch val data. The two `.backward()` calls in the file (lines 1049 and 1118) operate on training batches only (`warmup_loss` and `loss` from `train_loader.next_batch()`). ### Check 3: LEGAL score-first TTT (PR #1413 pattern, is_last_chunk guard) **NOT PRESENT.** No TTT mechanism of any kind exists. This is a standard training loop with no test-time adaptation. ### Check 4: HOLD scored-region SLOT **NOT PRESENT.** No scored-region slot manipulation. ### Check 5: CLEAN pure neural **CONFIRMED.** The submission is a clean architectural improvement: - 11 transformer layers (up from 9) with MLP 3x expansion (line 82: `mlp_mult=3`) - LeakyReLU(0.5)^2 activation (lines 707-708: `F.leaky_relu(..., negative_slope=0.5)` then `.square()`) - EMA weight averaging with decay 0.997 (lines 1136-1140) - Late QAT via STE when LR scale < 0.15 (lines 1143-1147; `CastedLinear.forward` lines 595-602) - Int6 per-row quantization...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
